src/main/scala/rise/core/Builder.scala

```diff
@@ -0,0 +1,22 @@
+package rise.core
+
+import types._
+
+trait Builder {
+  def apply: TypedDSL.TDSL[Primitive] =
+    throw new Exception("apply method must be overridden")
+  def apply(e: TypedDSL.TDSL[Expr]): TypedDSL.TDSL[App] =
+    TypedDSL.app(TypedDSL.toTDSL(apply), e)
+  def apply(n: Nat): TypedDSL.TDSL[DepApp[NatKind]] =
+    TypedDSL.depApp[NatKind](TypedDSL.toTDSL(apply), n)
+  def apply(dt: DataType): TypedDSL.TDSL[DepApp[DataKind]] =
+    TypedDSL.depApp[DataKind](TypedDSL.toTDSL(apply), dt)
+  def apply(a: AddressSpace): TypedDSL.TDSL[DepApp[AddressSpaceKind]] =
+    TypedDSL.depApp[AddressSpaceKind](TypedDSL.toTDSL(apply), a)
+
+  def unapply(arg: Expr): Boolean =
+    throw new Exception("unapply method must be overridden")
```

Comment:
I find it confusing that `unapply` deconstructs an `Expr` while `apply` constructs a `TDSL[Expr]`, especially with the additional `primitive` function on the side.

Reply:
The `TDSL` performs type inference on construction using the `TDSL` auxiliary type.
This is what is constructed with the `apply` method.
The `unapply` method is used for pattern matching in the rewriting where the `TDSL` type has been converted into the standard `Expr` type (after type inference).

The design of the `TDSL` is inherently asymmetrical: it uses the `TDSL` type for making the type inference work, but afterward (when you want to match) this type will be gone again.
The `Builder` trait basically reflects this design.